bftcosi/bftcosi.go

```diff
@@ -1,798 +1,239 @@
+	if bft.FinalSignatureChan == nil {
+		return fmt.Errorf("no FinalSignatureChan")
 	}
-	bft.closingMutex.Unlock()
 
-	// Start prepare round
-	if err := bft.handleAnnouncement(<-bft.announceChan); err != nil {
-		return err
-	}
-	if !bft.IsLeaf() {
-		if err := bft.handleCommitmentPrepare(bft.commitChan); err != nil {
-			return err
-		}
-	}
-
-	// Start commit round
-	if err := bft.handleAnnouncement(<-bft.announceChan); err != nil {
+	// prepare phase (part 1)
+	log.Lvl3("Starting prepare phase")
+	prepProto, err := bft.initCosiProtocol(phasePrep)
+	if err != nil {
 		return err
 	}
-	if !bft.IsLeaf() {
-		if err := bft.handleCommitmentCommit(bft.commitChan); err != nil {
-			return err
-		}
-	}
 
-	// Finish the prepare round
-	if err := bft.handleChallengePrepare(<-bft.challengePrepareChan); err != nil {
+	err = prepProto.Start()
+	if err != nil {
 		return err
 	}
-	if !bft.IsLeaf() {
-		if err := bft.handleResponsePrepare(bft.responseChan); err != nil {
-			return err
-		}
-	}
 
-	// Finish the commit round
-	if err := bft.handleChallengeCommit(<-bft.challengeCommitChan); err != nil {
-		return err
-	}
-	if !bft.IsLeaf() {
-		if err := bft.handleResponseCommit(bft.responseChan); err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
-// Signature will generate the final signature, the output of the BFTCoSi
-// protocol.
-// The signature contains the commit round signature, with the message.
-// If the prepare phase failed, the signature will be nil and the Exceptions
-// will contain the exception from the prepare phase. It can be useful to see
-// which cosigners refused to sign (each exceptions contains the index of a
-// refusing-to-sign signer).
-// Expect this function to have an undefined behavior when called from a
-// non-root Node.
-func (bft *ProtocolBFTCoSi) Signature() *BFTSignature {
-	bftSig := &BFTSignature{
-		Sig:        bft.commit.Signature(),
-		Msg:        bft.Msg,
-		Exceptions: nil,
-	}
-	if bft.signRefusal {
-		bftSig.Sig = nil
-		bftSig.Exceptions = bft.tempExceptions
-	}
-
-	// This is a hack to include exceptions which are the result of offline
-	// nodes rather than nodes that refused to sign.
-	if bftSig.Exceptions == nil {
-		for _, ex := range bft.tempExceptions {
-			if ex.Commitment.Equal(bft.Suite().Point().Null()) {
-				bftSig.Exceptions = append(bftSig.Exceptions, ex)
-			}
-		}
-	}
-
-	return bftSig
-}
-
-// RegisterOnDone registers a callback to call when the bftcosi protocols has
-// really finished
-func (bft *ProtocolBFTCoSi) RegisterOnDone(fn func()) {
-	bft.onDone = fn
-}
-
-// RegisterOnSignatureDone register a callback to call when the bftcosi
-// protocol reached a signature on the block
-func (bft *ProtocolBFTCoSi) RegisterOnSignatureDone(fn func(*BFTSignature)) {
-	bft.onSignatureDone = fn
-}
-
-// Shutdown closes all channels in case we're done
-func (bft *ProtocolBFTCoSi) Shutdown() error {
-	defer func() {
-		// In case the channels were already closed
-		recover()
+	go func() {
```

Comment:
How do you know this goroutine will terminate?

Reply:
It terminates when the prepProto (an instance of cosi), which it eventually does after a timeout.